libs/jwt-tools/src/Jwt/Tools.hs

```diff
@@ -0,0 +1,188 @@
+  StdMethod ->
+  Word16 ->
+  Epoch ->
+  Epoch ->
+  PEMKeys ->
+  ExceptT DPoPTokenGenerationError m DPoPAccessToken
+generateDpopToken dpopProof cid nonce uri method maxSkewSecs maxExpiration now backendPubkeyBundle = do
+  dpopProofCStr <- toCStr dpopProof
+  uidCStr <- toCStr $ ciUser cid
+  cidCUShort <- case readHex @Word16 (cs $ client $ ciClient cid) of
+    [(a, "")] -> pure a
+    _ -> throwE ClientIdSyntaxError
+  domainCStr <- toCStr $ ciDomain cid
+  nonceCStr <- toCStr nonce
+  uriCStr <- toCStr uri
+  methodCStr <- liftIO $ newCString $ cs $ methodToBS method
+  backendPubkeyBundleCStr <- toCStr backendPubkeyBundle
+
+  let before =
+        createToken
+          dpopProofCStr
+          uidCStr
+          cidCUShort
+          domainCStr
+          nonceCStr
+          uriCStr
+          methodCStr
+          maxSkewSecs
+          (epochNumber maxExpiration)
+          (epochNumber now)
+          backendPubkeyBundleCStr
+
+  let mkAccessToken response = do
+        case response of
+          Nothing -> pure $ Left FfiError
+          Just r -> do
+            mErr <- getError r
+            print mErr
+            mToken <- getToken r
+            pure $ toResult mErr mToken
+
+  let free = maybe (pure ()) free_dpop_access_token
+
+  ExceptT $ liftIO $ bracket before free mkAccessToken
+  where
+    toResult :: Maybe Word8 -> Maybe String -> Either DPoPTokenGenerationError DPoPAccessToken
+    -- the only valid case is when the error=0 (meaning no error) and the token is not null
+    toResult Nothing (Just token) = Right $ DPoPAccessToken $ cs token
+    -- error=1 corresponds to an unknown error on FFI side
+    toResult (Just 1) _ = Left FfiError
+    -- error=2 corresponds to 'FfiError' on FFI side
+    toResult (Just 2) _ = Left FfiError
+    -- error=3 corresponds to 'ImplementationError' on FFI side
+    toResult (Just 3) _ = Left FfiError
+    toResult (Just 4) _ = Left DpopSyntaxError
+    toResult (Just 5) _ = Left DpopTypError
+    toResult (Just 6) _ = Left DpopUnsupportedAlgorithmError
+    toResult (Just 7) _ = Left DpopInvalidSignatureError
+    toResult (Just 8) _ = Left ClientIdMismatchError
+    toResult (Just 9) _ = Left BackendNonceMismatchError
+    toResult (Just 10) _ = Left HtuMismatchError
+    toResult (Just 11) _ = Left HtmMismatchError
+    toResult (Just 12) _ = Left MissingJtiError
+    toResult (Just 13) _ = Left MissingChallengeError
+    toResult (Just 14) _ = Left MissingIatError
+    toResult (Just 15) _ = Left IatError
+    toResult (Just 16) _ = Left MissingExpError
+    toResult (Just 17) _ = Left ExpMismatchError
+    toResult (Just 18) _ = Left ExpError
+    -- this should also not happen, but apparently something went wrong
+    toResult _ _ = Left FfiError
+
+    toCStr :: forall a m. (ToByteString a, MonadIO m) => a -> m CString
+    toCStr = liftIO . newCString . cs . toByteString'
+
+    methodToBS :: StdMethod -> ByteString
+    methodToBS = \case
+      GET -> "GET"
```

Comment:
Isn't this `encode . show`?

Reply:
Show instances are meant to produce syntactically correct Haskell expressions, and that is not what we want here, even though it happens to give us the same result. I'd rather have a `ToByteString` instance, which does not exist for `StdMethod`. So we would have to wrap it into a `newtype`, but then I think a function is just as good, here.